lib/engine-addon.js

```diff
@@ -245,6 +245,49 @@ const buildEngineStyleTree = memoize(function buildEngineStyleTree() {
   );
 });
 
+const recoverAddons = function recoverAddons(addon) {
+  if (addon._orginalAddons) {
+    addon.addons = addon._orginalAddons;
+  }
+
+  if (addon.addons.length > 0) {
+    addon.addons.forEach(recoverAddons)
+  }
+}
+
+const deeplyNonDuplicatedAddon = function deeplyNonDuplicatedAddon(hostAddons, addon, treeName) {
```

Comment:
could this function be unit tested? That will make it easier to review as different scenarios can easily be validated :).

Reply:
Add test for "deeplyNonDuplicatedAddon".